File: app/src/main/java/it/notes/ecosystem/ui/DiaryScreen.kt
package it.notes.ecosystem.ui

import android.graphics.Bitmap
import androidx.activity.compose.BackHandler
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import it.notes.ecosystem.NotesApplication
import it.notes.ecosystem.domain.*
import it.notes.ecosystem.domain.Collection
import kotlinx.coroutines.*
import java.time.LocalDate
import java.time.YearMonth
import java.time.format.DateTimeFormatter
import java.util.Locale
import java.util.UUID

private val diaryDateFormat=DateTimeFormatter.ofPattern("d MMMM yyyy",Locale.ITALIAN)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun DiaryScreen(notes:List<Note>,collections:List<Collection>,repository:NotesRepository,
    onOpen:(String)->Unit,onCreate:(String,String?,Boolean)->Unit,onOpenTask:(String)->Unit,
    onSection:(String)->Unit,onManageCollections:()->Unit) {
    var selected by rememberSaveable {mutableStateOf(LocalDate.now().toString())}
    var monthText by rememberSaveable {mutableStateOf(YearMonth.now().toString())}
    var mode by rememberSaveable {mutableStateOf("Calendario")}
    var bookId by rememberSaveable {mutableStateOf<String?>(null)}
    var jump by rememberSaveable {mutableStateOf(false)}
    var newBook by rememberSaveable {mutableStateOf(false)}
    var bookName by rememberSaveable {mutableStateOf("")}
    var newTaskId by rememberSaveable {mutableStateOf<String?>(null)}
    var taskDate by rememberSaveable {mutableStateOf("")}
    var busy by remember {mutableStateOf(false)}
    var error by remember {mutableStateOf<String?>(null)}
    val scope=rememberCoroutineScope()
    val day=LocalDate.parse(selected);val month=YearMonth.parse(monthText)
    val selectedBook=collections.firstOrNull {it.id==bookId}
    // If a book was deleted remotely, expose all memories; do not pass a stale foreign key.
    val effectiveBook=selectedBook?.id
    val indexKey=notes to effectiveBook
    val result by produceState<Pair<Pair<List<Note>,String?>,DiaryIndex>?>(null,indexKey) {
        value=indexKey to withContext(Dispatchers.Default){Diary.index(notes,effectiveBook)}
    }
    val index=result?.takeIf {it.first==indexKey}?.second
    var today by remember {mutableStateOf(LocalDate.now())}
    LaunchedEffect(Unit) {while(true){today=LocalDate.now();delay(60000)}}
    BackHandler {if(!busy)onSection("Home")}
    fun action(block:suspend ()->Unit) {
        if(busy)return
        busy=true;error=null
        scope.launch {try {block()}catch(e:CancellationException){throw e}catch(e:Exception){error=e.message ?: "Operazione non riuscita."}finally{busy=false}}
    }
    Scaffold(topBar={TopAppBar(title={EditorialAppTitle("Diario")})},bottomBar={EcosystemNavigation("Diario",!busy,onSection)}) {padding->
        LazyColumn(Modifier.fillMaxSize().padding(padding),contentPadding=PaddingValues(20.dp),verticalArrangement=Arrangement.spacedBy(14.dp)) {
            item {EditorialSection("I giorni, le tue storie.","Pensieri, ricordi e impegni, ognuno al suo posto.")}
            item {Row(Modifier.horizontalScroll(rememberScrollState()),horizontalArrangement=Arrangement.spacedBy(8.dp)) {
                listOf("Calendario","Galleria","Book").forEach {label->FilterChip(mode==label,{mode=label},label={Text(label)})}
            }}
            if(error!=null) item {Text(error.orEmpty(),color=MaterialTheme.colorScheme.error)}
            if(busy || index==null) item {LinearProgressIndicator(Modifier.fillMaxWidth())}
            item {Row(Modifier.horizontalScroll(rememberScrollState()),horizontalArrangement=Arrangement.spacedBy(8.dp)) {
                FilterChip(effectiveBook==null,{bookId=null},label={Text("Tutti i ricordi")})
                collections.forEach {book->FilterChip(effectiveBook==book.id,{bookId=book.id},label={Text(book.name)})}
            }}
            if(mode=="Calendario") {
                item {Row(verticalAlignment=Alignment.CenterVertically) {
                    IconButton(enabled=month>YearMonth.of(1900,1),onClick={val next=month.minusMonths(1);monthText=next.toString();selected=next.atDay(minOf(day.dayOfMonth,next.lengthOfMonth())).toString()}){Icon(Icons.Default.ChevronLeft,"Mese precedente")}
                    TextButton(onClick={jump=true},modifier=Modifier.weight(1f)){Text(month.format(DateTimeFormatter.ofPattern("MMMM yyyy",Locale.ITALIAN)))}
                    IconButton(enabled=month<YearMonth.of(2200,12),onClick={val next=month.plusMonths(1);monthText=next.toString();selected=next.atDay(minOf(day.dayOfMonth,next.lengthOfMonth())).toString()}){Icon(Icons.Default.ChevronRight,"Mese successivo")}
                }}
                item {DiaryMonthGrid(month,day,today,index){selected=it.toString()}}
                item {Row(Modifier.horizontalScroll(rememberScrollState()),horizontalArrangement=Arrangement.spacedBy(14.dp)) {
                    listOf("Foto" to MaterialTheme.colorScheme.primary,"Pensieri" to MaterialTheme.colorScheme.secondary,"Impegni" to MaterialTheme.colorScheme.tertiary).forEach {(label,color)->
                        Row(verticalAlignment=Alignment.CenterVertically,horizontalArrangement=Arrangement.spacedBy(5.dp)) {Box(Modifier.size(8.dp).background(color,CircleShape));Text(label,style=MaterialTheme.typography.labelMedium)}
                    }
                }}
                item {TextButton(onClick={selected=today.toString();monthText=YearMonth.from(today).toString()}){Text("Torna a oggi")}}
                item {EditorialSection(day.format(diaryDateFormat),selectedBook?.let {"Ricordi in ${it.name} · impegni di tutti i book"})}
                item {Column(verticalArrangement=Arrangement.spacedBy(8.dp)) {
                    Button(onClick={onCreate(selected,effectiveBook,false)},modifier=Modifier.fillMaxWidth()){Text("Scrivi un pensiero")}
                    OutlinedButton(onClick={onCreate(selected,effectiveBook,true)},modifier=Modifier.fillMaxWidth()){Text("Aggiungi foto")}
                    TextButton(enabled=day.year in 2000..2200,onClick={taskDate=selected;newTaskId=UUID.randomUUID().toString()}){Text("Aggiungi impegno")}
                }}
                val entries=index?.days?.get(day).orEmpty()
                val tasks=index?.tasks?.get(day).orEmpty()
                if(index!=null && entries.isEmpty() && tasks.isEmpty()) item {Text("Questo giorno aspetta la sua prima storia.")}
                items(entries,key={"entry:${it.note.id}"}) {DiaryEntryCard(it,onOpen)}
                if(tasks.isNotEmpty()) item {Text("Impegni",style=MaterialTheme.typography.titleMedium)}
                items(tasks,key={"task:${it.id}"}) {task->
                    OutlinedCard(onClick={onOpenTask(task.id)},modifier=Modifier.fillMaxWidth()) {
                        Column(Modifier.padding(16.dp)) {
                            Text(task.title,style=MaterialTheme.typography.titleMedium)
                            Text(if(task.task?.completedAt!=null) "Completato" else "Da fare",style=MaterialTheme.typography.labelMedium)
                        }
                    }
                }
            } else if(mode=="Book" && effectiveBook==null) {
                item {Text("Un book usa una raccolta: le note datate diventano le sue pagine. Le altre note della raccolta restano nella libreria.")}
                item {Button(enabled=!busy,onClick={newBook=true}){Text("Crea book")};TextButton(onClick=onManageCollections){Text("Gestisci nomi e raccolte")}}
                if(collections.isEmpty()) item {Text("Vacanze, noi due, un anno da ricordare: scegli il nome del primo book.")}
                items(collections,key={"book:${it.id}"}) {book->
                    val entries=index?.entries.orEmpty().filter {it.note.collectionId==book.id}
                    OutlinedCard(onClick={bookId=book.id},modifier=Modifier.fillMaxWidth()) {
                        Column(Modifier.padding(18.dp),verticalArrangement=Arrangement.spacedBy(8.dp)) {
                            entries.firstOrNull {it.photos.isNotEmpty()}?.photos?.firstOrNull()?.let {DiaryPhoto(it)}
                            Text(book.name,style=MaterialTheme.typography.headlineSmall)
                            Text("${entries.size} pagine · ${entries.sumOf {it.photos.size}} foto")
                        }
                    }
                }
            } else {
                item {EditorialSection(selectedBook?.name ?: "La tua galleria",if(mode=="Galleria") "I ricordi con foto, dal più recente." else "Le pagine del book, dal più recente.")}
                if(mode=="Book") item {
                    OutlinedButton(onClick={mode="Calendario"}){Text("Aggiungi una pagina: scegli il giorno")}
                    TextButton(onClick=onManageCollections){Text("Gestisci book")}
                }
                val entries=index?.entries.orEmpty().filter {mode!="Galleria" || it.photos.isNotEmpty()}
                if(index!=null && entries.isEmpty()) item {Text("Nessun ricordo qui. Apri il calendario per aggiungerne uno o assegna Giorno e Raccolta a una nota esistente.")}
                items(entries,key={"gallery:${it.note.id}"}) {DiaryEntryCard(it,onOpen)}
            }
            item {Text("I ricordi archiviati restano nel Diario. Il cestino e i modelli sono esclusi. Le bozze compaiono dopo Salva.",style=MaterialTheme.typography.bodySmall)}
        }
    }
    if(jump) DiaryDateDialog(day,onDismiss={jump=false},allowRemove=false,onDate={date->
        if(date!=null){selected=date.toString();monthText=YearMonth.from(date).toString()};jump=false
    })
    if(newBook) AlertDialog(onDismissRequest={if(!busy)newBook=false},title={Text("Nuovo book")},text={Column(Modifier.verticalScroll(rememberScrollState())) {
        OutlinedTextField(bookName,{bookName=it},label={Text("Nome del book")},singleLine=true,enabled=!busy)
        Text("Sarà anche una raccolta, accessibile da Note.")
        error?.let {Text(it,color=MaterialTheme.colorScheme.error)}
    }},confirmButton={TextButton(enabled=!busy && bookName.isNotBlank(),onClick={action {repository.createCollection(bookName);bookName="";newBook=false}}){Text("Crea")}},dismissButton={TextButton(enabled=!busy,onClick={newBook=false}){Text("Annulla")}})
    newTaskId?.let {id->TaskEditorDialog(id,null,notes,busy,error,onDismiss={if(!busy)newTaskId=null},onSave={title,body,details,_->
        action {repository.savePlannedTask(null,id,title,body,details);newTaskId=null}
    },initialDue=taskDate)}
}

@Composable
private fun DiaryMonthGrid(month:YearMonth,selected:LocalDate,today:LocalDate,index:DiaryIndex?,onSelect:(LocalDate)->Unit) {
    val cells=remember(month){Diary.monthCells(month)}
    BoxWithConstraints(Modifier.fillMaxWidth()) {
        val gridWidth=maxOf(maxWidth,364.dp)
        Box(Modifier.horizontalScroll(rememberScrollState())) {
            Column(Modifier.width(gridWidth)) {
                Row {listOf("L","M","M","G","V","S","D").forEach {Text(it,Modifier.weight(1f),textAlign=androidx.compose.ui.text.style.TextAlign.Center)}}
                cells.chunked(7).forEach {week->Row {
                    week.forEach {date->
                        val entries=date?.let {index?.days?.get(it)}.orEmpty();val tasks=date?.let {index?.tasks?.get(it)}.orEmpty()
                        val photos=entries.any {it.photos.isNotEmpty()};val thoughts=entries.any {it.photos.isEmpty()}
                        val chosen=date==selected
                        Surface(modifier=Modifier.weight(1f).padding(2.dp).heightIn(min=60.dp).then(if(date!=null) Modifier.clickable(role=androidx.compose.ui.semantics.Role.Button) {onSelect(date)}.semantics {contentDescription="${date.format(diaryDateFormat)}, ${entries.size} ricordi, ${tasks.size} impegni${if(chosen) ", selezionato" else ""}"} else Modifier),
                            shape=MaterialTheme.shapes.small,color=when {chosen->MaterialTheme.colorScheme.primaryContainer;date==today->MaterialTheme.colorScheme.secondaryContainer;else->MaterialTheme.colorScheme.surface}) {
                            Column(Modifier.padding(vertical=8.dp),horizontalAlignment=Alignment.CenterHorizontally,verticalArrangement=Arrangement.spacedBy(6.dp)) {
                                Text(date?.dayOfMonth?.toString().orEmpty(),style=MaterialTheme.typography.bodyMedium)
                                Row(horizontalArrangement=Arrangement.spacedBy(3.dp)) {
                                    if(photos) Box(Modifier.size(6.dp).background(MaterialTheme.colorScheme.primary,CircleShape))
                                    if(thoughts) Box(Modifier.size(6.dp).background(MaterialTheme.colorScheme.secondary,CircleShape))
                                    if(tasks.isNotEmpty()) Box(Modifier.size(6.dp).background(MaterialTheme.colorScheme.tertiary,CircleShape))
                                }
                            }
                        }
                    }
                }}
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun DiaryEntryCard(entry:DiaryEntry,onOpen:(String)->Unit) {
    OutlinedCard(onClick={onOpen(entry.note.id)},modifier=Modifier.fillMaxWidth()) {
        Column(Modifier.padding(16.dp),verticalArrangement=Arrangement.spacedBy(8.dp)) {
            entry.photos.firstOrNull()?.let {DiaryPhoto(it)}
            Text(entry.date.format(diaryDateFormat),style=MaterialTheme.typography.labelMedium)
            Text(entry.note.title.ifBlank {"Un ricordo"},style=MaterialTheme.typography.titleLarge)
            val preview=entry.note.body.take(600).replace(Regex("\\[[^\\]]*]\\(notes-asset://[^)]*\\)"),"").trim().take(180)
            if(preview.isNotBlank())Text(preview,maxLines=3)
            if(entry.photos.isNotEmpty())Text("${entry.photos.size} foto · Apri la pagina",style=MaterialTheme.typography.labelMedium)
        }
    }
}

@Composable
private fun DiaryPhoto(ref:AttachmentRef) {
    val app=LocalContext.current.applicationContext as NotesApplication
    val sync by app.githubSync.status.collectAsStateWithLifecycle()
    val bitmap by produceState<Bitmap?>(null,ref.key,sync.busy) {
        value=try {withContext(Dispatchers.IO){decodeAttachmentImage(app.attachments.verifiedFile(ref.key),640)}}
        catch(e:CancellationException){throw e}catch(_:Exception){null}
    }
    val image=bitmap
    if(image!=null) Image(image.asImageBitmap(),ref.name,Modifier.fillMaxWidth().height(180.dp),contentScale=ContentScale.Crop)
    else Surface(Modifier.fillMaxWidth().height(100.dp),color=MaterialTheme.colorScheme.surfaceContainer) {
        Box(contentAlignment=Alignment.Center){Text("Foto da caricare · apri Allegati")}
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
internal fun DiaryHomeCard(notes:List<Note>,onOpen:()->Unit) {
    var today by remember {mutableStateOf(LocalDate.now())}
    LaunchedEffect(Unit){while(true){today=LocalDate.now();delay(60000)}}
    val count=remember(notes,today){notes.count {Diary.noteDate(it)==today}}
    OutlinedCard(onClick=onOpen,modifier=Modifier.fillMaxWidth()) {
        Column(Modifier.padding(16.dp),verticalArrangement=Arrangement.spacedBy(4.dp)) {
            Text("La tua giornata",style=MaterialTheme.typography.titleLarge)
            Text(if(count==0) "Conserva un pensiero o una foto di oggi." else "$count ricordi di oggi · Apri il Diario")
        }
    }
}

@Composable
internal fun DiaryDateEditor(model:EditorViewModel) {
    var open by rememberSaveable {mutableStateOf(false)}
    val date=Diary.date(model.tags)
    val enabled=model.available && !model.saving && !model.mediaBusy && !model.checklistWorking
    TextButton(enabled=enabled,onClick={open=true}) {Text(date?.let {"Giorno · ${it.format(diaryDateFormat)}"} ?: "Giorno")}
    if(open) DiaryDateDialog(date,onDismiss={open=false},externalError=model.error,onDate={if(model.assignDiaryDate(it))open=false})
}

@Composable
private fun DiaryDateDialog(date:LocalDate?,onDismiss:()->Unit,allowRemove:Boolean=true,externalError:String?=null,onDate:(LocalDate?)->Unit) {
    var value by rememberSaveable {mutableStateOf((date ?: LocalDate.now()).toString())}
    var error by remember {mutableStateOf<String?>(null)}
    AlertDialog(onDismissRequest=onDismiss,title={Text("Scegli il giorno")},text={Column(Modifier.verticalScroll(rememberScrollState())) {
        Text("La data del ricordo è indipendente da quando hai creato la nota.")
        OutlinedTextField(value,{value=it},singleLine=true,label={Text("AAAA-MM-GG")})
        TextButton(onClick={value=LocalDate.now().toString()}){Text("Oggi")}
        if(allowRemove)TextButton(onClick={onDate(null)}){Text("Rimuovi dal calendario")}
        (error ?: externalError)?.let {Text(it,color=MaterialTheme.colorScheme.error)}
    }},confirmButton={TextButton(onClick={try {onDate(Diary.validDate(value.trim()))}catch(e:Exception){error=e.message ?: "Data non valida."}}){Text("Conferma")}},dismissButton={TextButton(onClick=onDismiss){Text("Annulla")}})
}

File: flutter_app/lib/src/screens/diary_screen.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

import '../domain/diary.dart';
import '../domain/note.dart';
import '../domain/planner.dart';
import '../widgets/editorial.dart';

enum _DiaryMode { calendar, gallery, book }

class DiaryScreen extends StatefulWidget {
  const DiaryScreen({
    required this.notes,
    required this.collections,
    required this.onOpen,
    required this.onCreate,
    required this.onOpenTask,
    required this.onCreateCollection,
    super.key,
  });

  final List<Note> notes;
  final List<NoteCollection> collections;
  final ValueChanged<Note> onOpen;
  final Future<void> Function(DateTime date, String? collectionId) onCreate;
  final ValueChanged<Note> onOpenTask;
  final Future<void> Function(String name) onCreateCollection;

  @override
  State<DiaryScreen> createState() => _DiaryScreenState();
}

class _DiaryScreenState extends State<DiaryScreen> {
  DateTime _selected = dateOnly(DateTime.now());
  late DateTime _month = DateTime(_selected.year, _selected.month);
  _DiaryMode _mode = _DiaryMode.calendar;
  String? _collectionId;
  bool _busy = false;
  String? _error;

  NoteCollection? get _selectedCollection {
    for (final collection in widget.collections) {
      if (collection.id == _collectionId) return collection;
    }
    return null;
  }

  DiaryIndex get _index => Diary.index(
        widget.notes,
        collectionId: _selectedCollection?.id,
      );

  Future<void> _run(Future<void> Function() action) async {
    if (_busy) return;
    setState(() {
      _busy = true;
      _error = null;
    });
    try {
      await action();
    } catch (error) {
      if (mounted) {
        setState(
            () => _error = error.toString().replaceFirst('Exception: ', ''));
      }
    } finally {
      if (mounted) setState(() => _busy = false);
    }
  }

  Future<void> _createBook() async {
    final controller = TextEditingController();
    final name = await showDialog<String>(
      context: context,
      builder: (context) => AlertDialog(
        title: const Text('Nuovo book'),
        content: TextField(
          controller: controller,
          autofocus: true,
          decoration: const InputDecoration(labelText: 'Nome del book'),
        ),
        actions: [
          TextButton(
              onPressed: () => Navigator.pop(context),
              child: const Text('Annulla')),
          FilledButton(
            onPressed: () {
              final value = controller.text.trim();
              if (value.isNotEmpty) Navigator.pop(context, value);
            },
            child: const Text('Crea'),
          ),
        ],
      ),
    );
    controller.dispose();
    if (name == null || !mounted) return;
    await _run(() => widget.onCreateCollection(name));
  }

  @override
  Widget build(BuildContext context) {
    final index = _index;
    return ListView(
      padding: const EdgeInsets.fromLTRB(20, 12, 20, 104),
      children: [
        const EditorialSection(
          'I giorni, le tue storie.',
          detail: 'Pensieri, ricordi e impegni, ognuno al suo posto.',
        ),
        Wrap(
          spacing: 8,
          children: [
            ChoiceChip(
              selected: _mode == _DiaryMode.calendar,
              onSelected: _busy
                  ? null
                  : (_) => setState(() => _mode = _DiaryMode.calendar),
              label: const Text('Calendario'),
            ),
            ChoiceChip(
              selected: _mode == _DiaryMode.gallery,
              onSelected: _busy
                  ? null
                  : (_) => setState(() => _mode = _DiaryMode.gallery),
              label: const Text('Galleria'),
            ),
            ChoiceChip(
              selected: _mode == _DiaryMode.book,
              onSelected:
                  _busy ? null : (_) => setState(() => _mode = _DiaryMode.book),
              label: const Text('Book'),
            ),
          ],
        ),
        const SizedBox(height: 10),
        SingleChildScrollView(
          scrollDirection: Axis.horizontal,
          child: Row(
            children: [
              FilterChip(
                selected: _selectedCollection == null,
                onSelected:
                    _busy ? null : (_) => setState(() => _collectionId = null),
                label: const Text('Tutti i ricordi'),
              ),
              const SizedBox(width: 8),
              ...widget.collections.map((collection) => Padding(
                    padding: const EdgeInsets.only(right: 8),
                    child: FilterChip(
                      selected: _selectedCollection?.id == collection.id,
                      onSelected: _busy
                          ? null
                          : (_) =>
                              setState(() => _collectionId = collection.id),
                      label: Text(collection.name),
                    ),
                  )),
            ],
          ),
        ),
        if (_busy) ...[
          const SizedBox(height: 8),
          const LinearProgressIndicator(),
        ],
        if (_error != null) ...[
          const SizedBox(height: 8),
          Text(_error!,
              style: TextStyle(color: Theme.of(context).colorScheme.error)),
        ],
        const SizedBox(height: 12),
        if (_mode == _DiaryMode.calendar)
          _calendar(index)
        else if (_mode == _DiaryMode.book && _selectedCollection == null)
          _books(index)
        else
          _gallery(index),
      ],
    );
  }

  Widget _calendar(DiaryIndex index) {
    final dayKey = dateKey(_selected);
    final entries = index.days[dayKey] ?? const <DiaryEntry>[];
    final tasks = index.tasks[dayKey] ?? const <Note>[];

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Row(
          children: [
            IconButton(
              onPressed: _busy ? null : () => _moveMonth(-1),
              icon: const Icon(Icons.chevron_left),
            ),
            Expanded(
              child: TextButton(
                onPressed: _busy ? null : _pickDate,
                child: Text(
                  DateFormat('MMMM yyyy', 'it_IT').format(_month),
                  style: Theme.of(context).textTheme.titleMedium,
                ),
              ),
            ),
            IconButton(
              onPressed: _busy ? null : () => _moveMonth(1),
              icon: const Icon(Icons.chevron_right),
            ),
          ],
        ),
        _MonthGrid(
          month: _month,
          selected: _selected,
          today: dateOnly(DateTime.now()),
          index: index,
          onSelect: (date) => setState(() => _selected = date),
        ),
        const SizedBox(height: 8),
        Wrap(
          spacing: 14,
          children: const [
            _LegendDot(label: 'Ricordi', icon: Icons.circle),
            _LegendDot(label: 'Impegni', icon: Icons.check_circle),
          ],
        ),
        TextButton(
          onPressed: _busy
              ? null
              : () => setState(() {
                    _selected = dateOnly(DateTime.now());
                    _month = DateTime(_selected.year, _selected.month);
                  }),
          child: const Text('Torna a oggi'),
        ),
        EditorialSection(
          DateFormat('d MMMM yyyy', 'it_IT').format(_selected),
          detail: _selectedCollection == null
              ? null
              : 'Ricordi in ${_selectedCollection!.name} · impegni di tutti i book',
        ),
        FilledButton.icon(
          onPressed: _busy
              ? null
              : () => _run(
                  () => widget.onCreate(_selected, _selectedCollection?.id)),
          icon: const Icon(Icons.edit),
          label: const Text('Scrivi un pensiero'),
        ),
        const SizedBox(height: 12),
        if (entries.isEmpty && tasks.isEmpty)
          const Text('Questo giorno aspetta la sua prima storia.'),
        ...entries.map(_entryCard),
        if (tasks.isNotEmpty) ...[
          const SizedBox(height: 12),
          Text('Impegni', style: Theme.of(context).textTheme.titleMedium),
          const SizedBox(height: 6),
          ...tasks.map((task) {
            final details = TaskDetails.tryDecode(task.taskJson);
            return Card(
              child: ListTile(
                onTap: () => widget.onOpenTask(task),
                title: Text(task.title.isEmpty ? 'Attività' : task.title),
                subtitle:
                    Text(details?.completed == true ? 'Completato' : 'Da fare'),
                trailing: const Icon(Icons.chevron_right),
              ),
            );
          }),
        ],
      ],
    );
  }

  Widget _books(DiaryIndex index) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        const Text(
          'Un book usa una raccolta: le note datate diventano le sue pagine. Le altre note della raccolta restano nella libreria.',
        ),
        const SizedBox(height: 10),
        FilledButton.icon(
          onPressed: _busy ? null : _createBook,
          icon: const Icon(Icons.add),
          label: const Text('Crea book'),
        ),
        const SizedBox(height: 12),
        if (widget.collections.isEmpty)
          const Text(
              'Vacanze, noi due, un anno da ricordare: scegli il nome del primo book.')
        else
          ...widget.collections.map((collection) {
            final entries =
                Diary.index(widget.notes, collectionId: collection.id).entries;
            return Card(
              child: ListTile(
                onTap: () => setState(() => _collectionId = collection.id),
                leading: const Icon(Icons.book),
                title: Text(collection.name),
                subtitle: Text('${entries.length} pagine'),
                trailing: const Icon(Icons.chevron_right),
              ),
            );
          }),
      ],
    );
  }

  Widget _gallery(DiaryIndex index) {
    final entries = index.entries.where((entry) {
      if (_mode != _DiaryMode.gallery) return true;
      return _attachmentCount(entry.note.body) > 0;
    }).toList(growable: false);

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        EditorialSection(
          _selectedCollection?.name ?? 'La tua galleria',
          detail: _mode == _DiaryMode.gallery
              ? 'I ricordi con allegati immagine, dal più recente.'
              : 'Le pagine del book, dal più recente.',
        ),
        if (_mode == _DiaryMode.book)
          OutlinedButton.icon(
            onPressed: _busy
                ? null
                : () => setState(() => _mode = _DiaryMode.calendar),
            icon: const Icon(Icons.calendar_month),
            label: const Text('Aggiungi una pagina: scegli il giorno'),
          ),
        const SizedBox(height: 8),
        if (entries.isEmpty)
          const Text(
              'Nessun ricordo qui. Apri il calendario per aggiungerne uno.')
        else
          ...entries.map(_entryCard),
      ],
    );
  }

  Widget _entryCard(DiaryEntry entry) {
    final preview = entry.note.body
        .replaceAll(RegExp(r'\[[^\]]*\]\(notes-asset://[^)]*\)'), '')
        .replaceAll(RegExp(r'\s+'), ' ')
        .trim();
    final attachments = _attachmentCount(entry.note.body);

    return Padding(
      padding: const EdgeInsets.only(bottom: 10),
      child: Card(
        child: InkWell(
          onTap: () => widget.onOpen(entry.note),
          child: Padding(
            padding: const EdgeInsets.all(16),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  DateFormat('d MMMM yyyy', 'it_IT').format(entry.date),
                  style: Theme.of(context).textTheme.labelMedium,
                ),
                const SizedBox(height: 4),
                Text(
                  entry.note.title.isEmpty ? 'Un ricordo' : entry.note.title,
                  style: Theme.of(context).textTheme.titleLarge,
                ),
                if (preview.isNotEmpty) ...[
                  const SizedBox(height: 6),
                  Text(preview, maxLines: 3, overflow: TextOverflow.ellipsis),
                ],
                if (attachments > 0) ...[
                  const SizedBox(height: 8),
                  Text('$attachments allegati immagine · Apri la pagina',
                      style: Theme.of(context).textTheme.labelMedium),
                ],
              ],
            ),
          ),
        ),
      ),
    );
  }

  void _moveMonth(int delta) {
    final next = DateTime(_month.year, _month.month + delta);
    final maxDay = DateTime(next.year, next.month + 1, 0).day;
    setState(() {
      _month = next;
      _selected = DateTime(
          next.year, next.month, _selected.day.clamp(1, maxDay).toInt());
    });
  }

  Future<void> _pickDate() async {
    final picked = await showDatePicker(
      context: context,
      initialDate: _selected,
      firstDate: DateTime(1900),
      lastDate: DateTime(2200, 12, 31),
    );
    if (picked == null || !mounted) return;
    setState(() {
      _selected = dateOnly(picked);
      _month = DateTime(picked.year, picked.month);
    });
  }
}

class _MonthGrid extends StatelessWidget {
  const _MonthGrid({
    required this.month,
    required this.selected,
    required this.today,
    required this.index,
    required this.onSelect,
  });

  final DateTime month;
  final DateTime selected;
  final DateTime today;
  final DiaryIndex index;
  final ValueChanged<DateTime> onSelect;

  @override
  Widget build(BuildContext context) {
    final cells = Diary.monthCells(month);
    return Column(
      children: [
        Row(
          children: const ['L', 'M', 'M', 'G', 'V', 'S', 'D']
              .map((label) => Expanded(child: Center(child: Text(label))))
              .toList(),
        ),
        const SizedBox(height: 4),
        GridView.builder(
          shrinkWrap: true,
          physics: const NeverScrollableScrollPhysics(),
          itemCount: cells.length,
          gridDelegate: const SliverGridDelegateWithFixedCrossAxisCount(
            crossAxisCount: 7,
            mainAxisExtent: 62,
            crossAxisSpacing: 4,
            mainAxisSpacing: 4,
          ),
          itemBuilder: (_, i) {
            final date = cells[i];
            if (date == null) return const SizedBox.shrink();
            final entries = index.days[dateKey(date)] ?? const <DiaryEntry>[];
            final tasks = index.tasks[dateKey(date)] ?? const <Note>[];
            final chosen = dateKey(date) == dateKey(selected);
            final isToday = dateKey(date) == dateKey(today);
            return Material(
              color: chosen
                  ? Theme.of(context).colorScheme.primaryContainer
                  : isToday
                      ? Theme.of(context).colorScheme.secondaryContainer
                      : Theme.of(context).colorScheme.surface,
              borderRadius: BorderRadius.circular(12),
              child: InkWell(
                borderRadius: BorderRadius.circular(12),
                onTap: () => onSelect(date),
                child: Padding(
                  padding:
                      const EdgeInsets.symmetric(vertical: 7, horizontal: 4),
                  child: Column(
                    children: [
                      Text('${date.day}'),
                      const Spacer(),
                      Row(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          if (entries.isNotEmpty)
                            Icon(Icons.circle,
                                size: 7,
                                color: Theme.of(context).colorScheme.primary),
                          if (entries.isNotEmpty && tasks.isNotEmpty)
                            const SizedBox(width: 3),
                          if (tasks.isNotEmpty)
                            Icon(Icons.circle,
                                size: 7,
                                color: Theme.of(context).colorScheme.tertiary),
                        ],
                      ),
                    ],
                  ),
                ),
              ),
            );
          },
        ),
      ],
    );
  }
}

class _LegendDot extends StatelessWidget {
  const _LegendDot({required this.label, required this.icon});
  final String label;
  final IconData icon;

  @override
  Widget build(BuildContext context) => Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          Icon(icon, size: 9, color: Theme.of(context).colorScheme.primary),
          const SizedBox(width: 5),
          Text(label, style: Theme.of(context).textTheme.labelMedium),
        ],
      );
}

int _attachmentCount(String body) =>
    RegExp(r'notes-asset://[^)\s]+').allMatches(body).length;
